server/sonar-server/src/main/java/org/sonar/server/authentication/JwtTokenUpdater.java

```diff
@@ -0,0 +1,144 @@
+  private static final int SESSION_REFRESH_IN_MINUTES = 5;
+
+  private static final String RAILS_USER_ID_SESSION = "user_id";
+
+  private final System2 system2;
+  private final DbClient dbClient;
+  private final Server server;
+  private final JwtToken jwtToken;
+
+  public JwtTokenUpdater(System2 system2, DbClient dbClient, Server server, JwtToken jwtToken) {
+    this.jwtToken = jwtToken;
+    this.server = server;
+    this.dbClient = dbClient;
+    this.system2 = system2;
+  }
+
+  void createNewJwtToken(String userLogin, HttpServletResponse response) {
+    String token = jwtToken.encode(JwtToken.Jwt.builder()
+      .setUserLogin(userLogin)
+      .setExpirationTimeInSeconds(SESSION_TIMEOUT_IN_SECONDS)
+      .build());
+
+    LOG.trace("Create session for {}", userLogin);
+    response.addCookie(createCookie(JWT_COOKIE, token, SESSION_TIMEOUT_IN_SECONDS));
+  }
+
+  void validateJwtToken(HttpServletRequest request, HttpServletResponse response) {
+    Optional<Cookie> jwtCookie = findCookie(JWT_COOKIE, request);
+    if (jwtCookie.isPresent()) {
+      validateJwtToken(jwtCookie.get(), request, response);
+    }
+  }
+
+  private void validateJwtToken(Cookie jwtCookie, HttpServletRequest request, HttpServletResponse response) {
+    String value = jwtCookie.getValue();
+    checkNotNull(value, "JWT cookie is null");
+    Optional<Claims> claims = jwtToken.decode(value);
+    Optional<UserDto> user = claims.isPresent() ? getUser(claims.get().getSubject()) : Optional.empty();
+    if (claims.isPresent() && user.isPresent()) {
+      refreshSession(claims.get(), user.get(), request, response);
+    } else {
+      removeSession(request, response);
+    }
+  }
+
+  private void refreshSession(Claims token, UserDto user, HttpServletRequest request, HttpServletResponse response) {
+    String userLogin = token.getSubject();
+    LOG.trace("Validate session of {}", userLogin);
+    request.getSession().setAttribute(RAILS_USER_ID_SESSION, user.getId());
+
+    Date tokenCreationDatePlusFiveMinutes = DateUtils.addMinutes(token.getIssuedAt(), SESSION_REFRESH_IN_MINUTES);
+    if (new Date(system2.now()).after(tokenCreationDatePlusFiveMinutes)) {
+      LOG.trace("Create new session for {}", userLogin);
+      createNewJwtToken(user.getLogin(), response);
+    }
+  }
+
+  private void removeSession(HttpServletRequest request, HttpServletResponse response) {
+    LOG.trace("Remove session");
+    request.getSession().removeAttribute(RAILS_USER_ID_SESSION);
+    response.addCookie(createCookie(JWT_COOKIE, null, 0));
+  }
+
+  private Cookie createCookie(String name, @Nullable String value, int expiry) {
+    Cookie cookie = new Cookie(name, value);
+    cookie.setPath(server.getContextPath() + "/");
+    cookie.setSecure(server.isSecured());
+    cookie.setHttpOnly(true);
+    cookie.setMaxAge(expiry);
+    return cookie;
+  }
+
+  private static Optional<Cookie> findCookie(String cookieName, HttpServletRequest request) {
```

Comment:
This method should be used by CsrfVerifier too, for instance by extracting the method in a new class `CookieUtils`.
Note that CsrfVerifier is buggy as request.getCookies() may return null.